Reduce duplication in route

davidgoli opened the pull request:
This branch is motivated in anticipation of some upcoming bugfixes. The state of the route is difficult to reason about due to numerous duplications and minor inconsistencies between the initial load and page 2+ loads. This branch attempts to bring them into line with each other. No tests have been changed, because no behavior has been changed.

hhff commented on addon/mixins/route.js:
```diff
@@ -205,72 +194,98 @@ export default Ember.Mixin.create({
+    }
+
+    this.set(this.get('_modelPath') + '.reachedInfinity', true);
+    if(this.infinityModelLoaded) {
```
`this.infinityModelLoaded` is always going to be undefined - it's not used anywhere else.  We should use `_canLoadMore`

davidgoli replied:
we're checking `_canLoadMore` as the first line of this method, this merely copies the previous lines 242:247 and 257:259

hhff commented on addon/mixins/route.js:
```diff
@@ -164,31 +164,20 @@ export default Ember.Mixin.create({
+        let totalPages = infinityModel.get(this.get('totalPagesParam'));
         this.set('currentPage', startingPage);
         this.set('_totalPages', totalPages);
         infinityModel.set('reachedInfinity', !this.get('_canLoadMore'));
```
Feel like we could use `_finishLoad` here instead of setting this inline, right?

davidgoli replied:
`_finishLoad` is doing one other thing that I'm not sure applies here: ```this.set(this.get('_modelPath') + '.reachedInfinity', true);```

davidgoli replied:
another issue seems to be that I'm getting test failures if I call the `infinityModelLoaded` hook here (which is the other thing that `_finishLoad` does). That doesn't seem correct, but I don't want to inadvertently break anything here...

davidgoli replied:
nm I found a workaround for that

hhff commented on addon/mixins/route.js:
```diff
@@ -205,72 +194,98 @@ export default Ember.Mixin.create({
+    if (this.get('_loadingMore') || !this.get('_canLoadMore')) {
+      this._finishLoad();
+      return false;
     }
```
why do we call `this._finishLoad()` if `_loadingMore` is true?

davidgoli replied:
the original code branch was:

```javascript
if (!this.get('_loadingMore') && this.get('_canLoadMore')) {
  // load more
  ...
} else {
  ...
  // equivalent of finishLoad()
}
```

Here, I've inverted the condition and flattened:

```javascript
if (this.get('_loadingMore') || !this.get('_canLoadMore')) {
  this._finishLoad();
}

// load more
```

The logic _should_ be equivalent to before, just more easy to see what's going on. If that isn't correct, that may be out of scope.

davidgoli replied:
Perhaps the correct implementation is simply

```javascript
if (this.get('_loadingMore')) {
  return false;
}

if (!this.get('_canLoadMore')) {
  this._finishLoad();
  return false;
}

... // load more
```

hhff commented on addon/mixins/route.js:
```diff
@@ -205,72 +194,98 @@ export default Ember.Mixin.create({
+      let totalPages = this.get('_totalPages');
+      Ember.run.scheduleOnce('afterRender', this, 'infinityModelLoaded', { totalPages: totalPages });
+    }
+  },
```
Yeah - I mentioned it above but I feel like we should call `_finishLoad` (maybe `_loadingDidFinish` is a little more explicit) unconditionally at the end of every load.  That way we can do the short-circuit return here, and not wrap the method in any conditionals anywhere else

davidgoli replied:
I introduced a method `_notifyInfinityModelUpdated` above, perhaps making them consistent? Like `_modelDidUpdate` and `_modelDidFinish` ?

hhff commented:
Thankyou @davidgoli !  I've been putting this work off for a while now.  Thankyou a million!

davidgoli commented:
I'm traveling for the next week or so, so my response time may be a little slow, but I'll get to these issues ASAP.

davidgoli commented on addon/mixins/route.js:
```diff
@@ -205,72 +194,98 @@ export default Ember.Mixin.create({
+
+    promise.then(
+      newObjects => {
+        this.updateInfinityModel(newObjects);
```
in my other PR, all of this could go into `_notifyInfinityModelUpdated`

davidgoli commented on addon/mixins/route.js:
```diff
@@ -160,35 +172,26 @@ export default Ember.Mixin.create({
+    });
 
     if (typeof boundParams === 'object') {
       this.set('_boundParams', boundParams);
```
why the type check?

hhff replied:
probably should throw if not an object!

davidgoli commented on addon/mixins/route.js:
```diff
@@ -160,35 +172,24 @@ export default Ember.Mixin.create({
-
-    var requestPayloadBase = {};
-    requestPayloadBase[this.get('perPageParam')] = perPage;
-    requestPayloadBase[this.get('pageParam')] = startingPage;
```
I need to make sure the initial value of `startingPage` is being used in the first request. It's not as of this commit. I will also add a test so this failure is apparent.

davidgoli replied:
The problem before was that the initial/default value of `currentPage` wasn't used; instead `startingPage` was initialized like `var startingPage = options.startingPage || 1;` and then later on `this.set('currentPage', startingPage);`

davidgoli replied:
ok done

davidgoli commented:
Rewrote the route test to be more explicit about what failed when a failure happens. Also added an assertion to ensure the page number is correct (converted from `assert.ok(params.pageNumber)` to `assert.equals(params.pageNumber, expectedPageNumber)`

davidgoli commented:
It would be nice to standardize, when an asserted value is a boolean, to always use `assert.ok` or `assert.equals(value, true)`. Right now it's arbitrarily inconsistent.

davidgoli commented on addon/mixins/route.js:
```diff
@@ -160,117 +174,150 @@ export default Ember.Mixin.create({
+
+    this._loadNextPage();
+
+    return false;
```
So I don't think it's actually necessary to `return false;` from this method. The action itself does not return false, or return the value of this method, so it is probably unnecessary.

hhff replied:
personally, I think it should always return a promise

davidgoli commented:
@hhff OK, I think I'm done with the code churn here, pretty satisfied with how it's factored now. Please take a look at let me know if there's anything here you'd like me to change or clarify.

davidgoli commented:
In addition, this PR incorporates the code change in #89 and supercedes that branch.

hhff commented on addon/mixins/route.js:
```diff
@@ -102,6 +102,12 @@ export default Ember.Mixin.create({
 
+  actions: {
+    infinityLoad() {
+      this._infinityLoad();
```
lets return here

hhff commented on addon/mixins/route.js:
```diff
@@ -283,17 +278,61 @@ export default Ember.Mixin.create({
+   @method _notifyInfinityModelUpdated
+   */
+  _notifyInfinityModelUpdated(newObjects) {
+    if (!this.infinityModelUpdated) {
```
`this.infinityModelUpdated` is dead code, this is the only place it's referenced

hhff commented on addon/mixins/route.js:
```diff
@@ -283,17 +278,61 @@ export default Ember.Mixin.create({
+   @method _notifyInfinityModelLoaded
+   */
+  _notifyInfinityModelLoaded() {
+    if (!this.infinityModelLoaded) {
```
`this.infinityModelLoaded` is dead code too

davidgoli commented:
I thought `infinityModelUpdated` and `infinityModelLoaded` were supported hooks? They're in the documentation and probably also depended on in userland...

hhff commented:
oh yes they are sorry I forgot that Ember 2.0 requires hooks to be defined for them to run...  Ember 1.x did not - that's actually all good!

I need to find time to do one last look over all of this (this week sometime - this is a huge refactor so I wanna be certain we're good) but thankyou so so much for this amazing work @davidgoli

davidgoli commented:
@hhff merged in master

hhff commented on addon/mixins/route.js:
```diff
@@ -283,17 +278,61 @@ export default Ember.Mixin.create({
-   @return {Ember.Array} returns the updated infinity model
+   @return {DS.RecordArray} returns the updated infinity model
    */
   updateInfinityModel(newObjects) {
```
`updateInfinityModel` is a public hook that current addon consumers are using to modify what happens when a payload is received.  

By moving all of this functionality here, we're introducing breaking changes to users who are using this hook, and we're making the hook less useable.

Ember Infinity had over 250,000 downloads on NPM last month - this refactor is fantastic, but for such a simple lib we really can't be introducing breaking changes!

davidgoli replied:
Fair enough. I made this change because the goal is to call `updateInfinityModel` not just after page 2+ but also on page 1 (see https://github.com/hhff/ember-infinity/issues/90). I do think at a minimum we'll want to ensure `infinityModelUpdated` gets called.

On that topic, `updateInfinityModel` is a weird concept to me. What is it? Is it a hook? Is it a call to super? Particularly since this is a mixin, ideally the API would only use hooks that don't depend on a call to super, so users don't have to worry about return value or messing up functionality. The problem with the `infinityModelUpdated` hook however is that it's called outside of the current run loop, so that bound params may not be updated from new results before the next page load is triggered. We could leave `updateInfinityModel` and `infinityModelUpdated` as is, but we'd still need a hook that lets you update the bound params within the same run loop.

davidgoli replied:
I agree that we shouldn't introduce breaking changes in this PR, but I would almost advocate for deprecating the current usage of `updateInfinityModel`since it's inconsistent, and moving towards a new hook.

I'd also advocate for moving `currentPage` into the new `updateBoundParams` (or whatever) hook, and treating it as just another bound param. In our cursor-based infinity list, for example, the server doesn't even use the page number, it's just along for the ride.

davidgoli replied:
Not in this PR though of course

davidgoli replied:
Not sure why `updateInfinityModel` ever returned anything, the return value isn't used in calling code... but I'll maintain that functionality

davidgoli replied:
we should also add a test to ensure that we preserve the functionality you're concerned about... it's not currently tested

hhff commented:
Thanks @davidgoli - added a comment about a breaking change above.

I also love all the work you're doing refactoring the test suite!  That's going to close an issue we've got open for upping the code-climate.

However because it's such a huge refactor of the main codebase, I'd prefer we kept the old test suite for this PR, and move the new test suite refactor to a subsequent PR.  The old test suite hasn't let an regressions happen, and I'd be more comfortable merging this if it's passing.

davidgoli commented:
@hhff re: the test suite refactor, sure thing, I did all of that work in its own commit so it should be easy to split that off into its own branch.

davidgoli commented:
@hhff Not sure what's going on with the build, looks like an issue with the ember-data remote? It's not a test failure. I suspect this is happening across all builds right now.

```
bower EINVRES       Request to https://bower.herokuapp.com/packages/ember-data failed with 503
The command "bower install" failed and exited with 1 during .
```

davidgoli commented:
@hhff comments addressed

davidgoli commented on addon/mixins/route.js:
```diff
@@ -283,17 +281,78 @@ export default Ember.Mixin.create({
+    }
 
-    return infinityModel.pushObjects(newObjects.get('content'));
+    return newObjects;
```
on this first page, this array becomes the infinityModel that is returned on subsequent calls

hhff commented:
Hi @davidgoli !  

@mkorfmann & I did a review of this yesterday - we think it's so so good and we're so thankful for your work here!

This is our only suggestion (for readability):

This also means the `updateInfinityModel` hooks stays *exactly* the same.

```
  updateInfinityModel(newObjects) {
    let infinityModel = this._infinityModel();
    return infinityModel.pushObjects(newObjects.get('content'));
  },
```

```
  _nextPageLoaded(newObjects) {
    const totalPages = newObjects.get(this.get('totalPagesParam'));
    this.set('_totalPages', totalPages);

    let infinityModel;
    if (this.get('_firstPageLoaded')) {
      infinityModel = newObjects;
      this.set('_firstPageLoaded', true);
    } else {
      infinityModel = this.updateInfinityModel(newObjects);
    }

    this._notifyInfinityModelUpdated(newObjects);

    const canLoadMore = this.get('_canLoadMore');
    infinityModel.set('reachedInfinity', !canLoadMore);

    if (!canLoadMore) {
      this._notifyInfinityModelLoaded();
    }

    return infinityModel;
  }
```

davidgoli commented:
@hhff The problem with that change is that it regresses the fix for https://github.com/hhff/ember-infinity/issues/90 (this is also the same as the fix in https://github.com/hhff/ember-infinity/pull/89). If we're not going to call `updateInfinityModel` on the first page, then we'll need some other hook to fix that bug.

davidgoli commented:
@hhff ping - thanks for taking the time to review this - just want to verify that making the change you are asking is a conscious decision not to fix #90. We can a) adopt the changes as I have them, b) add an additional hook, or c) make a conscious decision to leave the bug unfixed.

hhff commented:
Thanks @davidgoli !  I think an Ember.run trigger like `this._notifyInfinityModelInitialized` will solve all our problems :+1:

hhff commented:
@davidgoli - also maybe checkout #77 - would that solve your problem in #90 ?

We plan to add that shortly after your PR gets in :+1:

davidgoli commented:
@hhff I've made the revert as you requested, let's merge this PR then we can hash out fixes to #90 separately. I also made a documentation change so `updateInfinityModel`'s quirky behavior is documented. I think a PR like #77 (which is basically the new hook I was talking about) could also be the fix.

davidgoli commented:
Also FYI I have that test refactor branch ready to go as soon as this gets merged!

mkorfmann commented:
Good stuff! Just want to give a big :+1: for the work you're doing here @davidgoli :).

BTW: Will #77 help you resolving #90 and #89?

Peace out :)

davidgoli commented:
@mkorfmann I'm into discussing #77, I think it has some flaws but I think the general approach is ok. We can discuss on that PR though, is this one good to go?

mkorfmann commented:
@davidgoli All right! Thanks for submitting sum feedback on #77 :).

This PR must be good to go, I'm tempted too just merge it but I wan to get @hhff's absolution before.

hhff commented:
Sorry all - been travelling!

ok @davidgoli - please squash the commits and I'll merge & release!  this is `0.1.0` :)

davidgoli commented:
@hhff squashed, but there's gonna be an extra commit, a merge from master, to resolve conflicts - FYI

kellyselden commented:
@davidgoli You can rebase upstream master to get it back to one commit :smile:

hhff commented:
what @kellyselden said! <3

davidgoli commented:
@hhff got it... it's done, should be gtg

hhff commented:
will merge when it passes, you're a total monster @davidgoli

davidgoli commented:
thanks for the review @hhff ! got another one coming right up

mkorfmann commented:
:tada:

2015-10-20 23:13 GMT+02:00 Hugh Francis <notifications@github.com>:

> Merged #88 <https://github.com/hhff/ember-infinity/pull/88>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hhff/ember-infinity/pull/88#event-440841989>.
>
